test/modelmatrix.jl

```diff
@@ -11,12 +15,15 @@
     d[:x2] = [9:12;]
     d[:x3] = [13:16;]
     d[:x4] = [17:20;]
+    d[:x1p] = CategoricalArray(d[:x1])
 
+    d_orig = deepcopy(d)
+    
     x1 = [5.:8;]
     x2 = [9.:12;]
     x3 = [13.:16;]
     x4 = [17.:20;]
-    f = @formula(y ~ x1 + x2)
+    f = @formula(y ~ 1 + x1 + x2)
```

Comment:
What happens now when no intercept is specified? As discussed several times, and I think we shouldn't silently omit the intercept if it doesn't appear in the formula, as it would be a trap for many users.

Reply:
Right now with `@formula` you get exactly what you ask for.  I agree this could be a "gotcha" for users but I think whether to include an intercept term by default or not is best left to the individual modeling packages, or at least pushed up to the model fitting API, especially since even statistical models have different assumptions about whether an intercept column is needed or permitted.  This is another case where we really need to know where the formula is "going" (as @oxinabox suggests and I mentioned above).  I'm not exactly sure how you'd do this with a schema, but maybe it belongs in the fit methods (which we can acomodate, maybe with a warning)

Reply:
We already have a mechanism to specify that a model doesn't want an intercept (https://github.com/JuliaStats/StatsModels.jl/pull/47). Couldn't we keep using that? I'd rather discuss whether we want to change the behavior of the intercept in a separate issue/PR, as this one is going to be quite long already.